Automattic/jetpack
# Jetpack: Extend disconnection dialog component to handle multiple steps and accept more props

This PR extends the disconnect dialog component so that it can show more information before disconnecting and then provide a feedback survey following disconnection. The dialog prop set is expanded with the intent of re-using the component in different contexts.

#### Changes proposed in this Pull Request:
* Update the disconnect dialog component to have a multi-step disconnect flow:
1 : Disconnect screen - show optional information passed via props and allow user to disconnect
2: Disconnect confirmation - shows following successful disconnection
3: Optional Feedback survey  - allow user to provide optional feedback after disconnecting
4: Thank you step - shows after the user elects to submit feedback

* Add several additional props to make the disconnect modal more flexible in multiple contexts
* Notably, props added include `connectedPlugins` that will show a list of other plugins that are using the Jetpack connection, and `disconnectComponent` which allows for a component to be passed that will render on the first screen of the disconnect flow.

#### Jetpack product discussion
pbNhbs-1n2-p2
p9dueE-3NP-p2

#### Does this pull request change what data or activity we track or use?
This PR allows the user to submit an optional feedback survey. The survey response is connected to the user's ID.
Analytics events are recorded for how a user interacts with the disconnect flow ( opening the flow, what steps are viewed, submitting the survey ) - see a full list of these events at the bottom of the test plan on this PR.

#### Related:
#21421 #20914

#### Testing instructions:
* Check out this branch on your Jetpack development environment
* The updates in this PR are most readily observed when disconnecting via the Jetpack Backups plugin, make sure you have the Jetpack Backups plugin installed and that Jetpack is connected.
* If you have not already, build the scripts for the Jetpack Backup plugin with `jetpack build`
* In the wp-admin, navigate to the admin page for the Jetpack Backup plugin and click the "Disconnect" link in the Connection section
* You should be presented with the first step of the disconnect flow, which will present a general warning message about disconnecting Jetpack:
![Screen Shot 2021-10-14 at 12 04 58 PM](https://user-images.githubusercontent.com/18016357/137379382-daac866c-39cd-48fc-b83f-4c701e7c721c.png)

* Click on "Disconnect" to disconnect the site. The button should show "Disconnecting" while disconnection is in progress.
![Screen Shot 2021-10-14 at 12 05 02 PM](https://user-images.githubusercontent.com/18016357/137379429-67e6507d-5a7b-4302-83a4-130a58cdbaad.png)

* After a successful disconnect, you will see the disconnect confirmation screen, which will show a "Back to my website" button. Clicking "Back to my website" will exit the disconnect flow and you should see the Jetpack connection screen for the Jetpack Backups plugin.
![Screen Shot 2021-11-20 at 10 38 24 AM](https://user-images.githubusercontent.com/18016357/142734135-54a30bfc-f2df-4cf2-b83f-84934b1a7671.png)

* This PR does include changes to show an additional survey step and a thank you step as part of the disconnection flow. However, it relies on the consumer of the JS connection package to provide some of the information required to collect the survey. Please see the test plan for #21421 to test the full disconnection flow, including the survey and thank you screens.


* This PR collects analytics events, but it does not have the correct scripts loaded to send them to tracks. Check that the `window._tkq` array contains events related to the disconnect flow. If you complete the test plan, you should see the following event names in `window.tkq` when on the "Jetpack Disconnected" step of the flow:
- `jetpack_disconnect_dialog_open`
- `jetpack_disconnect_dialog_step` with prop `step` of `disconnect`
- `jetpack_disconnect_dialog_step` with prop `step` of `disconnect_confirm`

Additional events that are collected in subsequent steps of the flow ( see test plan for #21421 ) are:
- `jetpack_disconnect_dialog_step` with prop `step` of `survey`
- `jetpack_disconnect_survey_submit`
- `jetpack_disconnect_dialog_step` with prop `step` of `thank_you`
